### PR TITLE
Reporting migrate - add support for configurable subjects

### PR DESCRIPTION
**Pull request by agorina**
What is not done: need to add 2 more cut_point(s) to asmt. Will do it next.

### CODE REVIEW

File: exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/SubjectRepositoryIT.java [hunk 1]
```diff
@@ -21,9 +21,9 @@
 @CachingTest
 @ActiveProfiles("test")
 @Sql(statements = {
-        "INSERT INTO subject (id, code) VALUES (-99, 'test1');",
-        "INSERT INTO subject (id, code) VALUES (-88, 'test2');",
-        "INSERT INTO subject (id, code) VALUES (-55, 'test55');"
+        "INSERT INTO subject (id, code, import_id, update_import_id) VALUES (-99, 'test1', 1, 1);",
```
Comment (agorina) [inline exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/SubjectRepositoryIT.java]: These two columns have been changed to `NOT NULL`, hence this change.

File: migrate-reporting/src/main/java/org/opentestsystem/rdw/ingest/migrate/reporting/MigrateReportingConfiguration.java [hunk 2]
```diff
@@ -27,16 +25,11 @@
     private static final Logger logger = LoggerFactory.getLogger(MigrateReportingConfiguration.class);
 
     private final JobBuilderFactory jobBuilderFactory;
-    private final StepBuilderFactory stepBuilderFactory;
-
-    public static final String stepWarehouseCheckName = "stepWarehouseCheck";
```
Comment (agorina) [inline migrate-reporting/src/main/java/org/opentestsystem/rdw/ingest/migrate/reporting/MigrateReportingConfiguration.java]: `stepBuilderFactory` was not used.

`stepWarehouseCheck` seems not needed with the configurable subjects. This step was verifying that the subjects' related data are in synch between reporting and warehouse.

File: migrate-reporting/src/test/java/org/opentestsystem/rdw/ingest/migrate/reporting/repository/impl/JdbcSqlListExecutionRepositoryIT.java [hunk 3]
```diff
@@ -20,15 +20,15 @@
     @Autowired
     private JdbcTemplate reportingJdbcTemplate;
 
-    @Sql(statements = "INSERT INTO item_trait_score (id, dimension) VALUES (-1, 'test1')")
```
Comment (agorina) [inline migrate-reporting/src/test/java/org/opentestsystem/rdw/ingest/migrate/reporting/repository/impl/JdbcSqlListExecutionRepositoryIT.java]: `item_trait_score` was removed from reporting. That is why this was changed.

File: migrate-reporting/src/test/resources/ReportingAllCodesSetup.sql [hunk 4]
```diff
@@ -6,38 +6,4 @@ INSERT INTO ethnicity(id, code) VALUES (-99,'AmericanIndianOrAlaskaNative-test')
 INSERT INTO gender(id, code) VALUES (-99,'Female-test'),(-98,'Male-test');
 INSERT INTO accommodation(id, code) VALUES(-99, 'code1-test'),( -98 ,'code2-test');
 INSERT INTO school_year(year) VALUES(1999),(1998);
-INSERT INTO depth_of_knowledge(id, level, subject_id, description, reference) VALUES (-99,1,1,'Recall and Reproduction','something'),(-98,2,1,'Basic Skills and Concepts','anything');
```
Comment (agorina) [inline migrate-reporting/src/test/resources/ReportingAllCodesSetup.sql]: It was moved and adjusted, these tables are not considered as 'code`s anymore.

File: migrate-reporting/src/test/resources/ReportingAndStagingEntitiesTeardown.sql [hunk 5]
```diff
@@ -65,4 +74,18 @@ DELETE FROM asmt_score where asmt_id < 0;
 DELETE FROM asmt_target where asmt_id < 0;
 DELETE FROM asmt where id < 0;
 
+-- ------------------------------------------ Subjects and related data ---------------------------------------------------------------------------------
+DELETE FROM depth_of_knowledge WHERE subject_id < 0 or id < 0;
+DELETE FROM common_core_standard  WHERE subject_id < 0 or id < 0;
+DELETE FROM subject_translation WHERE subject_id < 0 or label_code like '%integration test%';
+DELETE FROM subject_asmt_type WHERE subject_id < 0;
+DELETE FROM subject_claim_score WHERE subject_id < 0;
+-- TODO: this is not ideal since we are deleting more data that created by the test
```
Comment (agorina) [inline migrate-reporting/src/test/resources/ReportingAndStagingEntitiesTeardown.sql]: Yeah, this was my main struggle...

**Comment by wtritch** [comment]
Only 43 files changed?  Sure you didn't miss anything? =P

File: migrate-reporting/src/main/java/org/opentestsystem/rdw/ingest/migrate/reporting/step/StagingToReportingStepsConfig.java [hunk 6]
```diff
@@ -29,6 +29,7 @@
     static final String deleteCodesStepName = "deleteCodesStep";
     static final String deleteEntitiesStepName = "deleteEntitiesStep";
     static final String upsertOrganizationsStepName = "upsertOrganizationsStep";
+    static final String upsertSubjectsStepStepName = "upsertSubjectsStep";
```
Comment (jtreuting) [inline migrate-reporting/src/main/java/org/opentestsystem/rdw/ingest/migrate/reporting/step/StagingToReportingStepsConfig.java]: nit: double Step

File: migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml [hunk 7]
```diff
@@ -174,17 +174,124 @@ sql:
...
+                sscs.asmt_type_id,
+                sscs.code,
+                sscs.display_order,
+                sscs.data_order,
+                sscs.name
+              FROM staging_subject_claim_score sscs
+                LEFT JOIN subject_claim_score rscs ON rscs.id = sscs.id
+              WHERE rscs.id IS NULL;
+
+#         NOTE: updating data order is only possible when there is assessment or exam data loaded
```
Comment (wtritch) [inline migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml]: nit: "when there are no associated assessments or exams loaded"

File: migrate-reporting/src/main/java/org/opentestsystem/rdw/ingest/migrate/reporting/step/StagingToReportingStepsConfig.java (hunk 6 above)
Comment (wtritch) [reply migrate-reporting/src/main/java/org/opentestsystem/rdw/ingest/migrate/reporting/step/StagingToReportingStepsConfig.java]: I didn't know @agorina liked Dubstep

Comment (malaffoon) [reply migrate-reporting/src/main/java/org/opentestsystem/rdw/ingest/migrate/reporting/step/StagingToReportingStepsConfig.java]: Steppe steps step Steppe steps?

File: migrate-reporting/src/main/java/org/opentestsystem/rdw/ingest/migrate/reporting/step/StagingToReportingStepsConfig.java [hunk 8]
```diff
@@ -81,6 +82,39 @@ public Step deleteEntitiesStep() {
...
+                .addNext("claim", "insert")
+                .addNext("claim", "update")
+                .addNext("target", "insert")
+                .addNext("target", "update")
+                .addNext("common_core_standard", "deleteAsPartOfParentUpdate")
+                .addNext("common_core_standard", "insert")
+                .addNext("common_core_standard", "update")
+                .addNext("depth_of_knowledge", "deleteAsPartOfParentUpdate")
+                .addNext("depth_of_knowledge", "insert")
+                .addNext("depth_of_knowledge", "update");
```
Comment (malaffoon) [inline migrate-reporting/src/main/java/org/opentestsystem/rdw/ingest/migrate/reporting/step/StagingToReportingStepsConfig.java]: That's a lo of steps ... what could possibly go wrong?

File: migrate-reporting/src/main/java/org/opentestsystem/rdw/ingest/migrate/reporting/step/StagingToReportingStepsConfig.java (hunk 6 above)
Comment (jtreuting) [reply migrate-reporting/src/main/java/org/opentestsystem/rdw/ingest/migrate/reporting/step/StagingToReportingStepsConfig.java]: steps' ?

File: migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml [hunk 9]
```diff
@@ -174,17 +174,124 @@ sql:
...
+          update: >-
+            UPDATE subject rs
+              JOIN staging_subject ss ON ss.id = rs.id
+            SET
+              rs.updated = ss.updated,
+              rs.update_import_id = ss.update_import_id,
+              rs.migrate_id = ss.migrate_id
+
+        # ------------ Subject Asmt Type ------------------------------------------------------------------------
+      subject_asmt_type:
```
Comment (malaffoon) [inline migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml]: nit: indent

File: migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml [hunk 10]
```diff
@@ -174,17 +174,124 @@ sql:
...
+                ss.migrate_id
+              FROM staging_subject ss
+                LEFT JOIN subject rs ON rs.id = ss.id
+               WHERE rs.id IS NULL;
+
+          update: >-
+            UPDATE subject rs
+              JOIN staging_subject ss ON ss.id = rs.id
+            SET
+              rs.updated = ss.updated,
```
Comment (malaffoon) [inline migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml]: just checking: code isn't allowed to change?

File: migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml [hunk 11]
```diff
@@ -174,17 +174,124 @@ sql:
...
+             rsat.performance_level_standard_cutoff = ssat.performance_level_standard_cutoff,
+             rsat.claim_score_performance_level_count = ssat.claim_score_performance_level_count
+
+          deleteAsPartOfParentUpdate: >-
+            DELETE rsat FROM subject_asmt_type rsat
+              JOIN staging_subject ss ON ss.id = rsat.subject_id
+              WHERE NOT EXISTS( SELECT 1 FROM staging_subject_asmt_type ssat WHERE ssat.asmt_type_id = rsat.asmt_type_id and ssat.subject_id = rsat.subject_id);
+
+     # ---------------- subject_claim_score  -----------------------------------------------
+      subject_claim_score:
```
Comment (malaffoon) [inline migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml]: nit: indent

File: migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml (hunk 10 above)
Comment (wtritch) [reply migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml]: Nope, if you change the code and re-submit the subject xml you just created a new subject.

File: migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml [hunk 12]
```diff
@@ -174,17 +174,124 @@ sql:
...
+                sscs.code,
+                sscs.display_order,
+                sscs.data_order,
+                sscs.name
+              FROM staging_subject_claim_score sscs
+                LEFT JOIN subject_claim_score rscs ON rscs.id = sscs.id
+              WHERE rscs.id IS NULL;
+
+#         NOTE: updating data order is only possible when there is assessment or exam data loaded
+#         It is assumed that ingest manages this rule
```
Comment (malaffoon) [inline migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml]: nit: commenting is inconsistent (first column vs 8th or whatever)

File: migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml [hunk 13]
```diff
@@ -576,36 +640,48 @@ sql:
...
                               ,s.scale_score
                               ,s.scale_score_std_err
                               ,s.category
                             FROM staging_exam_claim_score s
-                              INNER JOIN exam_claim_score_mapping m ON m.subject_claim_score_id = s.subject_claim_score_id
-                                                                                 AND m.num = 4
+                              INNER JOIN subject_claim_score m ON m.id = s.subject_claim_score_id AND m.data_order = 4
                           ) AS claim4 ON claim4.exam_id = se.id
+                LEFT JOIN (
+                            SELECT s.exam_id
```
Comment (malaffoon) [inline migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml]: i know you're just copy and pasting but i'd vote for 670-673 being one line (and the other occurrences)

File: migrate-reporting/src/main/java/org/opentestsystem/rdw/ingest/migrate/reporting/step/StagingToReportingStepsConfig.java (hunk 6 above)
Comment (agorina) [reply migrate-reporting/src/main/java/org/opentestsystem/rdw/ingest/migrate/reporting/step/StagingToReportingStepsConfig.java]: That is a looooong discussion for this subject.... If this is the worst issue in this PR - than it is a great shape! :)

File: migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml (hunk 12 above)
Comment (agorina) [reply migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml]: Fixing. Trying to control my OCD to not reformat all these files 😉

File: migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml (hunk 13 above)
Comment (agorina) [reply migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml]: Changed.

**Pull request merged**